File: bot.py
import os
import logging
import requests
from telegram import Update, InlineKeyboardButton, InlineKeyboardMarkup
from telegram.ext import Updater, CommandHandler, CallbackQueryHandler, CallbackContext

# === Config ===
BOT_TOKEN = os.getenv("BOT_TOKEN")
REGISTER_API = "https://trumptossleaderboard-production.up.railway.app/register"
GAME_SHORT_NAME = "TrumpToss"
GAME_URL = "https://oscurantismo.github.io/trumptoss/"

if not BOT_TOKEN:
    print("❌ BOT_TOKEN not found in environment!")
    exit()
else:
    print("✅ BOT_TOKEN loaded.")

# === Logging ===
logging.basicConfig(
    format="%(asctime)s - %(name)s - %(levelname)s - %(message)s",
    level=logging.INFO
)
logger = logging.getLogger(__name__)

# === Helper ===
def register_user(username: str):
    try:
        response = requests.post(REGISTER_API, json={"username": username})
        logger.info(f"📨 Registered {username} → {response.status_code} – {response.json()}")
    except Exception as e:
        logger.error(f"❌ Failed to register {username}: {e}")

# === /start command ===
def start(update: Update, context: CallbackContext):
    user = update.effective_user
    username = user.username or f"user_{user.id}"
    logger.info(f"💬 /start by {username} (ID: {user.id})")

    register_user(username)

    keyboard = [
        [InlineKeyboardButton("👉 Play TrumpToss", callback_game={"game_short_name": GAME_SHORT_NAME})]
    ]
    reply_markup = InlineKeyboardMarkup(keyboard)

    context.bot.send_game(
        chat_id=update.effective_chat.id,
        game_short_name=GAME_SHORT_NAME,
        reply_markup=reply_markup
    )

# === Game Callback ===
def game_callback(update: Update, context: CallbackContext):
    query = update.callback_query
    logger.info(f"🎮 Game launch request: {query.game_short_name}")

    if query.game_short_name == GAME_SHORT_NAME:
        # ✅ Correctly respond with game URL so Telegram launches it
        context.bot.answer_callback_query(
            callback_query_id=query.id,
            url=GAME_URL
        )
        logger.info("✅ Game launched via answerCallbackQuery")
    else:
        context.bot.answer_callback_query(
            callback_query_id=query.id,
            text="Unknown game 🤔"
        )
        logger.warning("⚠️ Game short name did not match")

# === /status command ===
def status(update: Update, context: CallbackContext):
    update.message.reply_text("✅ TrumpToss bot is online and running!")

# === Optional bot status description ===
def set_bot_status(bot):
    try:
        bot.set_my_description("🟢 Online – TrumpToss bot is running!")
        logger.info("📍 Bot description updated")
    except Exception as e:
        logger.error(f"❌ Failed to set bot description: {e}")

# === Error Logger ===
def error_handler(update, context):
    logger.error(f"❌ Error: {context.error}")
    if update:
        logger.warning(f"⚠️ Caused by update: {update}")

# === Entry Point ===
def main():
    logger.info("🚀 Starting bot using long polling...")
    updater = Updater(BOT_TOKEN, use_context=True)
    dp = updater.dispatcher

    set_bot_status(updater.bot)

    dp.add_handler(CommandHandler("start", start))
    dp.add_handler(CommandHandler("status", status))
    dp.add_handler(CallbackQueryHandler(game_callback))
    dp.add_error_handler(error_handler)

    updater.start_polling()
    updater.idle()

if __name__ == "__main__":
    main()

File: bot.js
import { Bot, InlineKeyboard } from "grammy";
import dotenv from "dotenv";
import axios from "axios";

dotenv.config();

const bot = new Bot(process.env.BOT_TOKEN); // Place your bot token in .env

const GAME_SHORT_NAME = "TrumpToss";
const GAME_URL = "https://oscurantismo.github.io/trumptoss/";
const REGISTER_API = "https://trumptossleaderboard-production.up.railway.app/register";

// Helper: Register user
async function registerUser(id, first_name, last_name, username) {
    try {
        const fullUsername = username || `user_${id}`;
        const res = await axios.post(REGISTER_API, { username: fullUsername });
        console.log(`✅ Registered ${fullUsername} → ${res.data.status}`);
        return fullUsername;
    } catch (err) {
        console.error("❌ Registration error:", err.message);
        return null;
    }
}

// === /start ===
bot.command("start", async (ctx) => {
    const from = ctx.from;
    const name = await registerUser(from.id, from.first_name, from.last_name, from.username);

    await ctx.replyWithPhoto(
        "https://upload.wikimedia.org/wikipedia/commons/5/5f/President_Donald_Trump.jpg",
        {
            caption: `<b>Hi, ${from.first_name}!</b>\nPlay TrumpToss and climb the leaderboard!`,
            parse_mode: "HTML"
        }
    );
});

// === /play ===
bot.command("play", async (ctx) => {
    await registerUser(ctx.from.id, ctx.from.first_name, ctx.from.last_name, ctx.from.username);

    const keyboard = new InlineKeyboard()
        .game("🎮 Play TrumpToss")
        .row()
        .text("📊 Leaderboard", "leaderboard")
        .text("ℹ️ About", "about");

    await ctx.replyWithGame(GAME_SHORT_NAME, {
        reply_markup: keyboard,
        protect_content: true,
        disable_notification: true
    });
});

// === /leaderboard ===
bot.command("leaderboard", async (ctx) => {
    await ctx.reply("🏆 Leaderboard:\n1. Player1\n2. Player2\n3. Player3");
});

// === /about ===
bot.command("about", async (ctx) => {
    await ctx.reply("👟 TrumpToss is a Telegram game where you throw a shoe at Trump to score points.");
});

// === /help ===
bot.command("help", async (ctx) => {
    await ctx.reply(
        "<b>TrumpToss Help</b>\nThrow a shoe, score points, and hit the top leaderboard!\nContact @mora_dev for support.",
        { parse_mode: "HTML" }
    );
});

// === Inline Button Callbacks ===
bot.on("callback_query:data", async (ctx) => {
    const data = ctx.callbackQuery.data;
    if (data === "leaderboard") {
        await ctx.answerCallbackQuery();
        await ctx.reply("🏆 Leaderboard:\n1. Player1\n2. Player2\n3. Player3");
    } else if (data === "about") {
        await ctx.answerCallbackQuery();
        await ctx.reply("ℹ️ TrumpToss is just for fun.\nMade with love by @mora_dev.");
    }
});

// === Game Launch Callback ===
bot.on("callback_query:game_short_name", async (ctx) => {
    if (ctx.callbackQuery.game_short_name === GAME_SHORT_NAME) {
        await ctx.answerCallbackQuery({
            url: GAME_URL
        });
    } else {
        await ctx.answerCallbackQuery({
            text: "Unknown game",
            show_alert: true
        });
    }
});

// === Error Handling ===
bot.catch((err) => {
    console.error("❌ Bot Error:", err);
});

// === Start Bot ===
bot.start();
